### PR TITLE
Add Lab Page

### PR DESCRIPTION
**Pull request by blink1073**
jupyter-lab

screen shot 2016-02-18 at 9 43 30 am
screen shot 2016-02-18 at 9 43 42 am
screen shot 2016-02-18 at 9 43 54 am
screen shot 2016-02-18 at 9 44 06 am

### CODE REVIEW

**Comment by ellisonbg** [comment]
Very cool!

Could you add a note about this in the top-level README and describe any build steps needed (npm run build, etc.).

Also, we should have @cameronoelsen think about where to put a "JupyterLab" button or link in the main dashboard.

**Comment by blink1073** [comment]
There should be no change to the top level README, this should get run as part of setup.py.

**Comment by ellisonbg** [comment]
Don't see any changes to the setup.py though? Did that already happen?

**Comment by blink1073** [comment]
Nope, that's why its a WIP :)

**Comment by ellisonbg** [comment]
Ahh, got it. Should we just merge, or let you keep WIPing this PR?

On Mon, Jan 18, 2016 at 2:43 PM, Steven Silvester <notifications@github.com>
wrote:

> Nope, that's why its a WIP :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/972#issuecomment-172673663>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by blink1073** [comment]
I'd say wait until I figure out `setup.py` and satisfy Travis CI.

**Comment by ellisonbg** [comment]
OK, sounds good.

On Mon, Jan 18, 2016 at 2:46 PM, Steven Silvester <notifications@github.com>
wrote:

> I'd say wait until I figure out setup.py and satisfy Travis CI.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/jupyter/notebook/pull/972#issuecomment-172674126>.
>



-- 
Brian E. Granger
Associate Professor of Physics and Data Science
Cal Poly State University, San Luis Obispo
@ellisonbg on Twitter and GitHub
bgranger@calpoly.edu and ellisonbg@gmail.com

**Comment by blink1073** [comment]
Okay, it looks like just random failures on the builds.  I finished `setup.py` related actions, removing the WIP tag.

File: notebook/notebookapp.py [hunk 1]
```diff
@@ -239,7 +239,8 @@ def init_handlers(self, settings):
         handlers.extend(load_handlers('services.nbconvert.handlers'))
         handlers.extend(load_handlers('services.kernelspecs.handlers'))
         handlers.extend(load_handlers('services.security.handlers'))
-        
+        handlers.extend(load_handlers('lab.handlers'))
```
Comment (blink1073) [inline notebook/notebookapp.py]: Relevant change to this file.

File: notebook/static/lab/plugin.js [hunk 2]
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) Jupyter Development Team.
```
Comment (blink1073) [inline notebook/static/lab/plugin.js]: This file and its css will be moved to an `application` plugin in `jupyter-js-plugins` soon, pending some updates to `phosphide`

Comment (blink1073) [inline notebook/static/lab/plugin.js]: Done.

**Comment by cameronoelsen** [comment]
@blink1073 This is starting to look so slick :clap:

**Comment by jasongrout** [comment]
This is awesome.  Can you rebase/recommit the notebook/notebookapp.py and setupbase.py files so that the irrelevant whitespace changes are not in the commit?

**Comment by blink1073** [comment]
Done.

**Comment by jdfreder** [comment]
\* Subscribed

**Comment by blink1073** [comment]
Now with fancy link!:

screen shot 2016-02-11 at 3 04 12 pm

**Comment by jdfreder** [comment]
I like the graphic!

**Comment by cameronoelsen** [comment]
@jdfreder Thanks :grinning:

**Comment by jdfreder** [comment]
> @jdfreder Thanks

I should have known!

**Comment by blink1073** [comment]
In regards to the test failure, I looked through the history and this is affecting other open PRs randomly as well.  The test finishes but the process does not exit.  I suspect there is a thread spawned that is not successfully closing, since when I run the tests locally on Python 3.4 it hangs for some time before closing (on OSX).

**Comment by takluyver** [comment]
You can reproduce a hang? I've been trying to debug that, but it's a total heisenbug - every time I try to investigate it, there's no bug at all.

If you hit Ctrl-C while it's hanging after the tests, do you get any traceback that might indicate where it's hanging?

**Comment by jdfreder** [comment]
@takluyver you can't find the location of the heisenbug because you know too much about it's momentum...

**Comment by blink1073** [comment]
@takluyver, unfortunately Ctrl+C just causes it to bail with no further output.  I am seeing this though (regardless):

```python
Error in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/Users/ssilvester/anaconda/lib/python3.4/site-packages/IPython/core/history.py", line 749, in writeout_cache
    self._writeout_input_cache(conn)
  File "/Users/ssilvester/anaconda/lib/python3.4/site-packages/IPython/core/history.py", line 733, in _writeout_input_cache
    (self.session_number,)+line)
sqlite3.ProgrammingError: SQLite objects created in a thread can only be used in that same thread.The object was created in thread id 123145307557888 and this is thread id 140735176081408
```

**Comment by takluyver** [comment]
That was a separate issue - it should be fixed by jupyter/nbconvert#229.

If you have time to debug it some more: I suspect that it's [this test machinery](https://github.com/jupyter/notebook/blob/master/notebook/tests/launchnotebook.py) that's at fault. Can you try putting some print statements in `teardown_class` and `wait_until_dead` to see if it's hanging somewhere in there?

**Comment by blink1073** [comment]
Heisenberg strikes again, it doesn't seem to want to hang up anymore...

**Comment by blink1073** [comment]
I applied that patch @takluyver, and the SQLite error is indeed gone.   Adding print statements didn't show anything, because everything is now working smoothly (with and and without the nbconvert patch).

**Comment by jasongrout** [comment]
I hate to ask this again, but can you rebase/recommit the notebook/notebookapp.py and setupbase.py files so that the irrelevant whitespace changes are not in the commit? IIRC, we decided a long time ago that we could correct whitespace as we went in lines we changed, but we wouldn't make lots of other whitespace fixes since large amounts of unnecessary whitespace changes masks important relevant changes.

**Comment by jdfreder** [comment]
> I hate to ask this again, but can you rebase/recommit the notebook/notebookapp.py and setupbase.py files so that the irrelevant whitespace changes are not in the commit? IIRC, we decided a long time ago that we could correct whitespace as we went in lines we changed, but we wouldn't make lots of other whitespace fixes since large amounts of unnecessary whitespace changes masks important relevant changes.

I second this, I know it's a pain but it will help avoid rebasing with things like https://github.com/jupyter/notebook/pull/1047  :(

**Comment by blink1073** [comment]
Updated and rebased.

**Comment by sccolbert** [comment]
@blink1073 can you post updated screenshots?

**Comment by blink1073** [comment]
Updated.

**Comment by jasongrout** [comment]
I see we now have two commits in the list that put back in whitespace that was removed by other commits.  What I meant by rebase your commits adding whitespace was to change the original commits to not have any extra whitespace changes, so we don't have a lot of whitespace churn in the list of commits. It might be easiest at this point to just collapse everything to one commit.  Can you collapse things so that there isn't whitespace churn?

**Comment by blink1073** [comment]
We've uncovered some issues related to installation, and bootstrap interfering by adding `*` selector attributes, I will update this afternoon.

**Comment by jasongrout** [comment]
Okay, thanks.

**Comment by blink1073** [comment]
Ready for review, I can't replicate the install issues, but `python setup.py install` and `pip install -e .` work.

**Comment by minrk** [comment]
This is remarkably simple (I like it!). I landed the webpack PR, so there's conflict on setupbase.py. What do you think should be done before we merge this one? File browser's working great, but the notebook UI has a very long way to go.

**Comment by blink1073** [comment]
The original intent AFAIK was to demonstrate everything *but* the notebook itself.  Perhaps an initial released version should show read-only notebooks?

**Comment by minrk** [comment]
I think that makes sense.

**Comment by blink1073** [comment]
Let's wait for #1139 and I'll rebase.

**Comment by blink1073** [comment]
I'll add a read-only notebook plugin to `jupyter-js-plugins`, and we'll use that on the lab page.

**Comment by minrk** [comment]
Perfect, thanks!

**Comment by blink1073** [comment]
@minrk, I made the changes (see updated screen cast above).

**Comment by minrk** [comment]
It works! It would be cool if we could get the top-level `build:watch` to run on both targets.

**Comment by jdfreder** [comment]
:tada:

**Pull request merged**